File: solutions/698_partition_to_k_equal_sum_subsets.rb
# Given an array of integers nums and a positive integer k, find whether it's possible to divide this array into k non-empty subsets whose sums are all equal.
# Example 1:
# Input: nums = [4, 3, 2, 3, 5, 2, 1], k = 4
# Output: True
# Explanation: It's possible to divide it into 4 subsets (5), (1, 4), (2,3), (2,3) with equal sums.
# Note:
# 1 <= k <= len(nums) <= 16.
# 0 < nums[i] < 10000.
#

# @param {Integer[]} nums
# @param {Integer} k
# @return {Boolean}
def can_partition_k_subsets(nums, k)
  return true if nums.empty?
  return false if nums.size < k || nums.reduce(:+) % k > 0
  seen = Hash.new(0)
  possible = lambda do |i, cursum, n, target|
    return true if n == 1
    return false if cursum > target
    return possible.call(0, 0, n - 1, target) if cursum == target

    (i..nums.size - 1).each do |j|
      next unless seen[j].zero?
      seen[j] = 1
      return true if possible.call(j, cursum + nums[j], n, target)
      seen[j] = 0
    end
    return false
  end
  possible.call(0, 0, k, nums.reduce(:+) / k)
end

nums = [4, 3, 2, 3, 5, 2, 1]
nums = [2, 2, 2, 2, 3, 4, 5]
k = 4
p can_partition_k_subsets(nums, k)

File: _site/solutions/698_partition_to_k_equal_sum_subsets.cpp
#include <algorithm>
#include <climits>
#include <iostream>
#include <map>
#include <stdio.h>
#include <unordered_map>
#include <vector>
/**
Given an array of integers nums and a positive integer k, find whether it's
possible to divide this array into k non-empty subsets whose sums are all equal.
Example 1:
Input: nums = [4, 3, 2, 3, 5, 2, 1], k = 4
Output: True
Explanation: It's possible to divide it into 4 subsets (5), (1, 4), (2,3), (2,3)
with equal sums.
Note:
1 <= k <= len(nums) <= 16.
0 < nums[i] < 10000.

 **/
using namespace std;

static int speed_up = []() {
  std::ios::sync_with_stdio(false);
  cin.tie(NULL);
  return 0;
}();

class Solution {
public:
  bool canPartitionKSubsets(vector<int> &nums, int k) {
    if (nums.empty())
      return false;
    int sum = accumulate(nums.begin(), nums.end(), 0);
    if (nums.size() < k || sum % k != 0)
      return false;
    vector<int> seen(nums.size(), 0);
    return possible(nums, seen, 0, k, 0, sum / k);
  }

  bool possible(std::vector<int> &nums, std::vector<int> &seen, int i, int k,
                int cursum, int target) {
  	cout << i << ", " << k << ", " << cursum << endl; 
    if (k == 1)
      return true;
    if (cursum == target)
      return possible(nums, seen, 0, k - 1, 0, target);
  	if (cursum > target)
  		return false; 
    for (int j = i; j < nums.size(); j++) {
      if (!seen[j]) {
        seen[j] = 1;
        if (possible(nums, seen, j + 1, k, cursum + nums[j], target))
          return true;
        seen[j] = 0;
      }
    }
    return false;
  }
};

int main() {
  Solution s;
  vector<int> nums = {1, 4, 3, 5, 7};
  sort(nums.begin(), nums.end(), greater<int>());
  for (int n : nums)
    cout << n << " ";
  cout << endl;
  nums = {4, 3, 2, 3, 5, 2, 1};
  cout << s.canPartitionKSubsets(nums, 4) << endl;
}
